### PR TITLE
feat(backend): api keys database setup

### PR DESCRIPTION
**Pull request by stanleynguyen**
## Problem

Closes [ticket](https://www.notion.so/opengov/Setting-up-api_keys-support-5af9e208c9834481a5bcce135b5ee405?pvs=4)

## Solution

- Create `api_keys` database table
- Implement double write to`api_keys` table

## Deployment Checklist

_Any tasks that need to be done on the environment before releasing of this PR (e.g. Setting environment variables, running migrations, etc). If there's no task to be done, put "N/A"_

- [ ] Run DB migration
- [ ] Granting privileges to backend IAM user on the new table and sequence

**Post-deployment**
- [ ] Run `migrate-api-keys` script

### CODE REVIEW

File: backend/src/migrate-api-keys.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import sequelizeLoader from '@core/loaders/sequelize.loader'
+import { User } from '@core/models'
+import { ApiKey } from '@core/models/user/api-key'
+import { Op } from 'sequelize'
+void (async function main() {
+  await sequelizeLoader()
+  const users = await User.findAll({
+    where: { apiKeyHash: { [Op.not]: null } },
+  })
+  for (const u of users) {
```
Comment (richard-yak) [inline backend/src/migrate-api-keys.ts:14]: can be optimized if performance is a concern, but if it's a one time script to be ran after hours then no issues

File: backend/src/database/migrations/20230221073046-create-api-keys-table.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+module.exports = {
+  up: async (queryInterface, Sequelize) => {
+    await queryInterface.createTable('api_keys', {
+      id: {
+        allowNull: false,
+        autoIncrement: true,
+        primaryKey: true,
+        type: Sequelize.BIGINT,
+      },
+      user_id: {
```
Comment (richard-yak) [inline backend/src/database/migrations/20230221073046-create-api-keys-table.js:12]: probably a worthwhile effort to create index on this col now before the user size api key table grows big.

also remember there was a set active / inactive functionality briefly discussed before, i'm guessing that's no longer required!

**Review by richard-yak — APPROVED** [review]
seemsGood, minor optimizations can be considered but likely irrelevant till the table gets huge

File: backend/src/migrate-api-keys.ts (hunk 1 above)
Comment (stanleynguyen) [reply backend/src/migrate-api-keys.ts:14]: Actually this was intended, cos we don't have a lot of user records to begin with (can be queried within a single query without limit) and we're narrowing down to `api_key_hash not null`

File: backend/src/database/migrations/20230221073046-create-api-keys-table.js (hunk 2 above)
Comment (stanleynguyen) [reply backend/src/database/migrations/20230221073046-create-api-keys-table.js:12]: - The indexing piece is interesting 🤔 we might actually need cos in the future we expiring keys
- Yea the expiry, but we're not doing it here yet. This is just for multi API keys first

File: backend/src/migrate-api-keys.ts (hunk 1 above)
Comment (tan-zx) [reply backend/src/migrate-api-keys.ts:14]: we have 46 active API users haha

Comment (richard-yak) [reply backend/src/migrate-api-keys.ts:14]: oh then defo no issue

**Comment by tan-zx** [comment]
Haven't looked through in detail, but I was thinking the rate-limit should be at the key level, rather than the user level. That way, ICA can have multiple sub-systems, each with their own rate limit.

Doesn't have to be in this PR, but I was thinking an RFC visualising the long-term state of the API key system would be helpful! Then we can decide how to implement these features in parts.

File: backend/src/migrate-api-keys.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
+// eslint-disable-next-line @typescript-eslint/no-var-requires
```
Comment (tan-zx) [inline backend/src/migrate-api-keys.ts:1]: just curious — this seems like a weird place to put this script, why don't we run it as a database migration and keep track of it with all the other migration scripts there?

**Review by tan-zx — APPROVED** [review]
small nit re: placement of migration script, otherwise lgtm

File: backend/src/migrate-api-keys.ts (hunk 3 above)
Comment (stanleynguyen) [reply backend/src/migrate-api-keys.ts:1]: Cos it's not a schema migration, there's no `down`, it's a one-way-street kinda activity

**Pull request merged**